File: Serenity.Script.UI/Editor/Select2AjaxEditor.cs
using jQueryApi;
using System;
using System.Collections.Generic;
using System.Html;
using System.Runtime.CompilerServices;

namespace Serenity
{
    [Imported(ObeysTypeSystem = true), Element("<input type=\"hidden\"/>")]
    [IncludeGenericArguments(false), ScriptName("Select2AjaxEditor")]
    public abstract class Select2AjaxEditor<TOptions, TItem> : Widget<TOptions>, IStringValue
        where TOptions : class, new()
        where TItem: class
    {
        protected int pageSize = 50;

        static Select2AjaxEditor()
        {
            Q.Prop(typeof(Select2AjaxEditor<object, object>), "value");
        }

        public Select2AjaxEditor(jQueryObject hidden, TOptions opt)
            : base(hidden, opt)
        {
            var emptyItemText = EmptyItemText();
            if (emptyItemText != null)
                hidden.Attribute("placeholder", emptyItemText);

            hidden.Select2(GetSelect2Options());

            hidden.Attribute("type", "text"); // jquery validate to work
            
            hidden.Bind2("change." + this.uniqueName, (e, x) =>
            {
                if (e.HasOriginalEvent() || Q.IsFalse(x))
                {
                    if (hidden.GetValidator() != null)
                        hidden.Valid();
                }
            });
        }

        protected virtual string EmptyItemText()
        {
            return element.GetAttribute("placeholder") ?? Q.Text("Controls.SelectEditor.EmptyItemText");
        }

        protected virtual string GetService()
        {
            throw new NotImplementedException();
        }

        protected virtual void Query(ListRequest request, Action<ListResponse<TItem>> callback)
        {
            var options = new ServiceCallOptions<ListResponse<TItem>>
            {
                BlockUI = false,
                Service = GetService() + "/List",
                Request = request,
                OnSuccess = response =>
                {
                    callback(response);
                }
            };

            ExecuteQuery(options);
        }

        protected virtual void ExecuteQuery(ServiceCallOptions<ListResponse<TItem>> options)
        {
            Q.ServiceCall(options);
        }

        protected virtual void QueryByKey(string key, Action<TItem> callback)
        {
            var options = new ServiceCallOptions<RetrieveResponse<TItem>>
            {
                BlockUI = false,
                Service = GetService() + "/Retrieve",
                Request = new RetrieveRequest
                {
                    EntityId = key.As<Int64>()
                },
                OnSuccess = response =>
                {
                    callback(response.Entity);
                }
            };

            ExecuteQueryByKey(options);
        }

        protected virtual void ExecuteQueryByKey(ServiceCallOptions<RetrieveResponse<TItem>> options)
        {
            Q.ServiceCall(options);
        }

        protected abstract string GetItemKey(TItem item);
        protected abstract string GetItemText(TItem item);

        protected virtual int GetTypeDelay()
        {
            return 500;
        }

        protected virtual Select2Options GetSelect2Options()
        {
            var emptyItemText = EmptyItemText();
            int queryTimeout = 0;

            return new Select2Options
            {
                MinimumResultsForSearch = 10,
                PlaceHolder = !emptyItemText.IsEmptyOrNull() ? emptyItemText : null,
                AllowClear = emptyItemText != null,
                Query = delegate(Select2QueryOptions query)
                {
                    var request = new ListRequest
                    {
                        ContainsText = query.Term.TrimToNull(),
                        Skip = (query.Page - 1) * pageSize,
                        Take = pageSize + 1
                    };

                    if (queryTimeout != 0)
                        Window.ClearTimeout(queryTimeout);

                    queryTimeout = Window.SetTimeout(delegate
                    {
                        Query(request, response =>
                        {
                            query.Callback(new Select2Result
                            {
                                Results = response.Entities.Slice(0, pageSize).As<List<TItem>>().Map(x => new Select2Item
                                {
                                    Id = GetItemKey(x),
                                    Text = GetItemText(x),
                                    Source = x
                                }),
                                More = response.Entities.Count >= pageSize
                            });
                        });
                    }, GetTypeDelay());
                },
                InitSelection = delegate(jQueryObject element, Action<object> callback)
                {
                    var val = element.GetValue();
                    if (val.IsEmptyOrNull())
                    {
                        callback(null);
                        return;
                    }

                    QueryByKey(val, result =>
                    {
                        callback(result == null ? null : new Select2Item
                        {
                            Id = GetItemKey(result),
                            Text = GetItemText(result),
                            Source = result
                        });
                    });
                }
            };
        }

        protected void AddInplaceCreate(string title)
        {
            var self = this;

            J("<a><b/></a>").AddClass("inplace-button inplace-create")
                .Attribute("title", title)
                .InsertAfter(this.element)
                .Click(e =>
                {
                    self.InplaceCreateClick(e);
                });

            this.Select2Container.Add(this.element).AddClass("has-inplace-button");
        }

        protected virtual void InplaceCreateClick(jQueryEvent e)
        {
        }

        protected jQueryObject Select2Container
        {
            get { return this.element.PrevAll(".select2-container"); }
        }

        public string Value
        {
            get
            {
                return this.element.Select2Get("val") as string;
            }
            set
            {
                if (value != Value)
                    this.element.Select2("val", value).TriggerHandler("change", new object[] { true });
            }
        }
    }
}

File: Serenity.Scripts/CoreLib/UI/Editors/Select2AjaxEditor.ts
﻿namespace Serenity {

    @Decorators.registerEditor('Serenity.Select2AjaxEditor', [IStringValue])
    @Decorators.element('<input type="hidden" />')
    export class Select2AjaxEditor<TOptions, TItem> extends Widget<TOptions> implements IStringValue {
        pageSize: number = 50;

        constructor(hidden: JQuery, opt: TOptions) {
            super(hidden, opt);

            var emptyItemText = this.emptyItemText();
            if (emptyItemText != null)
                hidden.attr("placeholder", emptyItemText);

            hidden.select2(this.getSelect2Options());

            hidden.attr("type", "text"); // jquery validate to work

            hidden.on("change." + this.uniqueName, (e, x) => {
                if (WX.hasOriginalEvent(e) || !x) {
                    if (ValidationHelper.getValidator(hidden) != null)
                        hidden.valid();
                }
            });
        }

        protected emptyItemText(): string {
            var txt = this.element.attr('placeholder');
            if (txt == null) {
                txt = Q.text('Controls.SelectEditor.EmptyItemText');
            }
            return txt;
        }

        protected getService(): string {
            throw new Error("Not implemented!");
        }

        protected query(request: ListRequest, callback: (p1: ListResponse<any>) => void): void {
            var options: Q.ServiceOptions<any> = {
                blockUI: false,
                service: this.getService() + '/List',
                request: request,
                onSuccess: function (response) {
                    callback(response);
                }
            };
            this.executeQuery(options);
        }

        protected executeQuery(options: ServiceOptions<ListResponse<any>>): void {
            Q.serviceCall(options);
        }

        protected queryByKey(key: string, callback: (p1: any) => void): void {
            var options: Q.ServiceOptions<any> = {
                blockUI: false,
                service: this.getService() + '/Retrieve',
                request: { EntityId: key },
                onSuccess: function (response) {
                    callback(response.Entity);
                }
            };
            this.executeQueryByKey(options);
        }

        protected executeQueryByKey(options: ServiceOptions<RetrieveResponse<any>>): void {
            Q.serviceCall(options);
        }

        protected getItemKey(item: any): string {
            return null;
        }

        protected getItemText(item: any): string {
            return null;
        }

        protected getTypeDelay(): number {
            return 500;
        }

        protected getSelect2Options(): Select2Options {
            var emptyItemText = this.emptyItemText();
            var queryTimeout = 0;
            return {
                minimumResultsForSearch: 10,
                placeHolder: (!Q.isEmptyOrNull(emptyItemText) ? emptyItemText : null),
                allowClear: Q.isValue(emptyItemText),
                query: query => {
                    var request = {
                        ContainsText: Q.trimToNull(query.term), Skip: (query.page - 1) * this.pageSize, Take: this.pageSize + 1
                    };

                    if (queryTimeout !== 0) {
                        window.clearTimeout(queryTimeout);
                    }

                    queryTimeout = window.setTimeout(() => {
                        this.query(request, response => {
                            query.callback({
                                results: response.Entities.slice(0, this.pageSize).map(x => {
                                    return { id: this.getItemKey(x), text: this.getItemText(x), source: x };
                                }), more: response.Entities.length >= this.pageSize
                            });
                        });
                    }, this.getTypeDelay());

                },
                initSelection: (element, callback) => {
                    var val = element.val();
                    if (Q.isEmptyOrNull(val)) {
                        callback(null);
                        return;
                    }
                    this.queryByKey(val, result => {
                        callback((result == null ? null : {
                            id: this.getItemKey(result),
                            text: this.getItemText(result),
                            source: result
                        }));
                    });
                }
            };
        }

        protected addInplaceCreate(title: string): void {
            var self = this;
            $('<a><b/></a>').addClass('inplace-button inplace-create')
                .attr('title', title).insertAfter(this.element).click(function (e) {
                    self.inplaceCreateClick(e);
                });

            this.get_select2Container().add(this.element)
                .addClass('has-inplace-button');
        }

        protected inplaceCreateClick(e: any): void {
        }

        protected get_select2Container(): JQuery {
            return this.element.prevAll('.select2-container');
        }

        get_value(): string {
            return Q.safeCast(this.element.select2('val'), String);
        }

        get value(): string {
            return this.get_value();
        }

        set_value(value: string) {
            if (value !== this.get_value()) {
                var el = this.element;
                el.select2('val', value);
                el.data('select2-change-triggered', true);
                try {
                    el.triggerHandler('change', [true]); // valueSet: true
                }
                finally {
                    el.data('select2-change-triggered', false);
                }
            }
        }

        set value(v: string) {
            this.set_value(v);
        }
    }
}